pkg/query/optimize_test.go

```diff
@@ -0,0 +1,470 @@
+		t.Parallel()
+
+		// Create a typed optimizer that only works on Union
+		typedOptimizer := func(u *Union) (Iterator, bool, error) {
+			return u, true, nil // Would return true if called
+		}
+
+		// Wrap it and use in ApplyOptimizations
+		wrapped := WrapOptimizer[*Union](typedOptimizer)
+
+		// Test with an Intersection - should not match
+		intersection := NewIntersection()
+		result, changed, err := ApplyOptimizations(intersection, []OptimizerFunc{wrapped})
+		require.NoError(t, err)
+		require.False(t, changed)
+		require.Equal(t, intersection, result)
+	})
+}
+
+func TestElideSingletonUnionAndIntersection(t *testing.T) {
+	t.Parallel()
+
+	t.Run("elides singleton union", func(t *testing.T) {
+		t.Parallel()
+
+		union := NewUnion()
+		fixed := newNonEmptyFixedIterator()
+		union.addSubIterator(fixed)
+
+		result, changed, err := ApplyOptimizations(union, []OptimizerFunc{ElideSingletonUnionAndIntersection})
+		require.NoError(t, err)
+		require.True(t, changed)
+		require.Equal(t, fixed, result)
+	})
+
+	t.Run("elides singleton intersection", func(t *testing.T) {
+		t.Parallel()
+
+		intersection := NewIntersection()
+		fixed := newNonEmptyFixedIterator()
+		intersection.addSubIterator(fixed)
+
+		result, changed, err := ApplyOptimizations(intersection, []OptimizerFunc{ElideSingletonUnionAndIntersection})
+		require.NoError(t, err)
+		require.True(t, changed)
+		require.Equal(t, fixed, result)
+	})
+
+	t.Run("does not elide multi-element union", func(t *testing.T) {
+		t.Parallel()
+
+		union := NewUnion()
+		union.addSubIterator(newNonEmptyFixedIterator())
+		union.addSubIterator(newNonEmptyFixedIterator())
+
+		result, changed, err := ApplyOptimizations(union, []OptimizerFunc{ElideSingletonUnionAndIntersection})
+		require.NoError(t, err)
+		require.False(t, changed)
+		require.Equal(t, union, result)
+	})
+
+	t.Run("does not elide multi-element intersection", func(t *testing.T) {
+		t.Parallel()
+
+		intersection := NewIntersection()
+		intersection.addSubIterator(newNonEmptyFixedIterator())
+		intersection.addSubIterator(newNonEmptyFixedIterator())
+
+		result, changed, err := ApplyOptimizations(intersection, []OptimizerFunc{ElideSingletonUnionAndIntersection})
+		require.NoError(t, err)
+		require.False(t, changed)
+		require.Equal(t, intersection, result)
+	})
+
+	t.Run("does not elide other iterator types", func(t *testing.T) {
+		t.Parallel()
+
+		fixed := newNonEmptyFixedIterator()
+		result, changed, err := ApplyOptimizations(fixed, []OptimizerFunc{ElideSingletonUnionAndIntersection})
+		require.NoError(t, err)
+		require.False(t, changed)
+		require.Equal(t, fixed, result)
+	})
+}
+
+func TestRemoveNullIterators(t *testing.T) {
+	t.Parallel()
+
+	t.Run("removes empty fixed from union", func(t *testing.T) {
+		t.Parallel()
+
+		union := NewUnion()
+		fixed := newNonEmptyFixedIterator()
+		empty := NewEmptyFixedIterator()
+		union.addSubIterator(fixed)
+		union.addSubIterator(empty)
```

Comment:
for a follow-up PR: this would be slightly easier to read if we could do

```go
fixed := newNonEmptyFixedIterator()
empty := NewEmptyFixedIterator()
union := NewUnion(fixed, empty)
```